lib/runtime/newWasmer/memory.go

```diff
@@ -0,0 +1,38 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package newWasmer
+
+import (
+	"errors"
+	"fmt"
+
+	wasmgo "github.com/wasmerio/wasmer-go/wasmer"
+)
+
+var errGrowMemory = errors.New("failed to grow memory")
+
+// Memory is a thin wrapper around Wasmer memory to support
+// Gossamer runtime.Memory interface
+type Memory struct {
+	memory *wasmgo.Memory
+}
+
+// Data returns the memory's data
+func (m Memory) Data() []byte {
+	return m.memory.Data()
+}
+
+// Length returns the memory's length
+func (m Memory) Length() uint32 {
+	return uint32(m.memory.DataSize())
```

Comment:
I'm not set on this, and am open to feedback (going to tag @kishansagathiya here) but I think we are okay. In particular I am hesitant change because the wasmer grow function takes in a pages type, which contains a receiver func `ToUint32()` so since the type already seems to operate using uint32, my instinct is to leave as is.

But also like I said, I'm open to more feedback on this

Reply:
Good question, it does not raise an exception however it saturates the values to the `uint32` boundaries: https://go.dev/play/p/xyeyeO9CmWE

Reply:
@EclesioMeloJunior do you think this is okay behavior for this situation?

Reply:
If I'm not wrong, using uint32 will limit the memory up to 4Gb. Are we doing it intentionally?

Reply:
Maybe my memory is wrong (overflow?), but isn't wasmer limited to 32 bit anyway? 🤔

On top of that, I have never seen 64 bit pointers in our entire wasmer calls.

Reply:
So I think it is not a problem to leave it as it is since, as @qdm12 says, currently wasm is limited to 32-bit aka `wasm32`, so by default in the current wasm model the limit is 4GB, more info: https://v8.dev/blog/4gb-wasm-memory

Reply:
Just to make sure I understand, to prevent unanticipated unsafe casts we will first cast to a uint64 and then do a bounds check on the value? This is desirable because we can have custom error handling in the case of an invalid cast and also we could realize if there is a change to wasm (e.i. move to uint64)

Is my understanding here correct @kishansagathiya?

Reply:
@kishansagathiya I made a change so that we still use uint32 with respect to wasmer, but with the additions adding a check to make sure that the uint is withing the proper bounds. Let me know if further modifications are needed!